### PR TITLE
types,expression: make ConvertToTime unaffected by location

### PR DESCRIPTION
**Pull request by tiancaiamao**
Fix issue https://github.com/pingcap/tidb/issues/5436

Convert Duration to Datetime needs to mix the time part with current date, but `ConvertToTime` use time.Local, which cause the result affected by system time zone.

Both Duration and Datetime have no location information, I change them to use UTC, and avoid converting to Go time (which involves location).

Some test code use local time zone, I changed them accordingly.

@XuHuaiyu

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
What will `select cast(time('12:23:34') as date);` return in different time zone?

**Comment by tiancaiamao** [comment]
Before fix, system timezone set to UTC+8:00

```
mysql> select cast(time('12:23:34') as date);
+--------------------------------+
| cast(time('12:23:34') as date) |
+--------------------------------+
| 2018-01-30                     |
+--------------------------------+
1 row in set (0.00 sec)
```

system timezone set to Denver:

```
mysql> select cast(time('12:23:34') as date);
+--------------------------------+
| cast(time('12:23:34') as date) |
+--------------------------------+
| 2018-01-29                     |
+--------------------------------+
1 row in set (0.00 sec)
```

After this PR:

```
mysql> select cast(time('12:23:34') as date);
+--------------------------------+
| cast(time('12:23:34') as date) |
+--------------------------------+
| 2018-01-30                     |
+--------------------------------+
1 row in set (0.00 sec)
```

**Comment by XuHuaiyu** [comment]
Is the following query result consistent with MySQL after this commit? @tiancaiamao 

In MySQL:
``` sql
mysql> set time_zone="-08:00"; 
Query OK, 0 rows affected (0.00 sec)

mysql> select cast(time('12:23:34') as date);
+--------------------------------+
| cast(time('12:23:34') as date) |
+--------------------------------+
| 2018-01-29                     |
+--------------------------------+
1 row in set (0.00 sec)

mysql> set time_zone="+08:00";
Query OK, 0 rows affected (0.00 sec)

mysql> select cast(time('12:23:34') as date);
+--------------------------------+
| cast(time('12:23:34') as date) |
+--------------------------------+
| 2018-01-30                     |
+--------------------------------+
1 row in set (0.00 sec)
```

File: types/time.go [hunk 1]
```diff
@@ -840,13 +840,14 @@ func (d Duration) ToNumber() *MyDecimal {
 // ConvertToTime converts duration to Time.
 // Tp is TypeDatetime, TypeTimestamp and TypeDate.
 func (d Duration) ConvertToTime(tp uint8) (Time, error) {
-	year, month, day := gotime.Now().Date()
-	// just use current year, month and day.
-	n := gotime.Date(year, month, day, 0, 0, 0, 0, gotime.Local)
```
Comment (winkyao) [inline types/time.go]: why not just `gotime.Date(year, month, day, 0, 0, 0, 0, gotime.UTC)` ?

Comment (tiancaiamao) [reply types/time.go]: Use `mixDateAndTime` instead of Go time library make it more MySQL compatible.

File: types/time.go [hunk 2]
```diff
@@ -840,13 +840,14 @@ func (d Duration) ToNumber() *MyDecimal {
...
-	n = n.Add(d.Duration)
+	year, month, day := gotime.Now().In(gotime.UTC).Date()
```
Comment (coocood) [inline types/time.go]: I think we should use the TimeZone in `SessionVars`.

**Comment by coocood** [comment]
LGTM

**Comment by tiancaiamao** [comment]
PTAL @XuHuaiyu  @winkyao

**Comment by XuHuaiyu** [comment]
add test for these cases mentioned before @tiancaiamao 
```
mysql> set time_zone="-08:00"; 
Query OK, 0 rows affected (0.00 sec)

mysql> select cast(time('12:23:34') as date);
+--------------------------------+
| cast(time('12:23:34') as date) |
+--------------------------------+
| 2018-01-29                     |
+--------------------------------+
1 row in set (0.00 sec)

mysql> set time_zone="+08:00";
Query OK, 0 rows affected (0.00 sec)

mysql> select cast(time('12:23:34') as date);
+--------------------------------+
| cast(time('12:23:34') as date) |
+--------------------------------+
| 2018-01-30                     |
+--------------------------------+
1 row in set (0.00 sec)
```

**Comment by tiancaiamao** [comment]
DONE. @XuHuaiyu

**Pull request merged**